media-gfx/openvdb/openvdb-11.0.0.ebuild

```diff
@@ -0,0 +1,247 @@
+# Copyright 1999-2023 Gentoo Authors
+# Distributed under the terms of the GNU General Public License v2
+
+EAPI=8
+
+PYTHON_COMPAT=( python3_{10..12} )
+
+LLVM_MAX_SLOT=14
+
+inherit cmake cuda llvm python-single-r1
+
+DESCRIPTION="Library for the efficient manipulation of volumetric data"
+HOMEPAGE="https://www.openvdb.org"
+SRC_URI="https://github.com/AcademySoftwareFoundation/${PN}/archive/v${PV}.tar.gz -> ${P}.tar.gz"
+
+LICENSE="MPL-2.0"
+SLOT="0/$PV"
+KEYWORDS="~amd64 ~arm ~arm64 ~ppc64 ~riscv ~x86"
+IUSE="+abi11-compat abi10-compat abi8-compat abi9-compat ax +blosc cpu_flags_x86_avx cpu_flags_x86_sse4_2
+	cuda doc +nanovdb numpy python static-libs test utils"
+RESTRICT="!test? ( test )"
+
+REQUIRED_USE="^^ ( abi8-compat abi9-compat abi10-compat abi11-compat )
+	cpu_flags_x86_avx? ( cpu_flags_x86_sse4_2 )
+	cuda? ( nanovdb )
+	numpy? ( python )
+	python? ( ${PYTHON_REQUIRED_USE} )"
+
+# utils? media-gfx/alembic not yet keyworded on ppc64
+RDEPEND="
+	>=dev-cpp/tbb-2020.3:=
+	dev-libs/boost:=
+	dev-libs/jemalloc:=
+	dev-libs/log4cplus:=
+	>=dev-libs/imath-3.1.4-r2:=
+	sys-libs/zlib:=
+	x11-libs/libXcursor
+	x11-libs/libXi
+	x11-libs/libXinerama
+	x11-libs/libXrandr
+	ax? ( <sys-devel/llvm-15:= )
+	blosc? ( dev-libs/c-blosc:= )
+	cuda? ( >=dev-util/nvidia-cuda-toolkit-11 )
+	python? (
+		${PYTHON_DEPS}
+		$(python_gen_cond_dep '
+			dev-libs/boost:=[numpy?,python?,${PYTHON_USEDEP}]
+			numpy? ( dev-python/numpy[${PYTHON_USEDEP}] )
+		')
+	)
+	utils? (
+		media-libs/glfw
+		media-libs/glu
+		media-libs/libjpeg-turbo:=
+		media-libs/libpng:=
+		>=media-libs/openexr-3:=
+		virtual/opengl
+	)
+"
+DEPEND="${RDEPEND}"
+BDEPEND="
+	virtual/pkgconfig
+	doc? (
+		app-doc/doxygen
+		dev-texlive/texlive-bibtexextra
+		dev-texlive/texlive-fontsextra
+		dev-texlive/texlive-fontutils
+		dev-texlive/texlive-latex
+		dev-texlive/texlive-latexextra
+	)
+	test? (
+		dev-cpp/gtest
+		dev-util/cppunit
+	)
+"
+
+PATCHES=(
+	"${FILESDIR}/${PN}-8.1.0-glfw-libdir.patch"
+	"${FILESDIR}/${PN}-9.0.0-fix-atomic.patch"
+	"${FILESDIR}/${PN}-10.0.1-fix-linking-of-vdb_tool-with-OpenEXR.patch"
+	"${FILESDIR}/${PN}-10.0.1-drop-failing-tests.patch"
+	"${FILESDIR}/${PN}-10.0.1-log4cplus-version.patch"
+	"${FILESDIR}/${PN}-11.0.0-constexpr-version.patch"
+	"${FILESDIR}/${PN}-11.0.0-NANOVDB_USE_OPENVDB.patch"
+)
+
+cuda_set_CUDAHOSTCXX() {
```

Comment:
Isn't these two cuda functions better suited to be in the cuda eclass?

I think it is probably better to have them in the eclass so other packages with cuda can easily adopt this logic as well.

There already seem to be a function for getting the highest compatible gcc version in the cuda eclass with `cuda_gccdir`. But I guess it had some pitfalls that you tried to workaround in your own version?

Reply:
They ideally would, but I am not fully certain what it will look in the end. Fixing things in an eclass also takes longer then in a package. I adjusted them as I fixed the cuda support for a bunch of packages. The functions and package adjustments are the result of my local overrides that I used to use cuda. So I don't have to explain in `#gentoo` over and over again how to get cuda working.

They are in my other pull requests as well. `cuda_gccdir` relies on `cuda-config` which has hard coded values. It only supports `gcc`. My version tests if the current compiler (`gcc` or `clang`) can compile cuda code and otherwise tries to find an older compiler version which should be installed as a result of the `nvidia-cuda-toolkit` dependencies.

`CUDAHOSTCXX` and `CUDAARCHS` are also corresponding cmake variables when cuda is used as first class language with cmake.

Reply:
If you have already added it to other ebuilds as well, wouldn't it be better to add the additional functions to the eclass so you don't have to maintain multiple implementation scattered around in different ebuilds?

As the existing `cuda_gccdir` is more restrictive than the never functions you wrote. I guess introducing new functions to the eclass would be best?

As the `CUDAHOSTCXX` and `CUDAARCHS` is cmake specific, they shouldn't be set directly with the eclass. But I think it would be trivial to assign the result of the functions to those variables, no?

Reply:
Eclass changes have to, as far as I am aware, go over gentoo-dev ml. Which takes a while. Eclasses should also, from my understanding, contain well tested code.

So putting this into an eclass now would mean I can't change it quickly and would have to be very certain that I don't have to change it as often as I might want to. I also don't have to test every ebuild affected by those changes.

Reply:
It wasn't too tedious for me when I added git-lfs support to the git eclass: https://github.com/gentoo/gentoo/pull/31905
As you can see I had the changes accepted within 10 days.
Of course this doesn't mean it will go as smooth for you.

But I think it is good to at least try.

Note that I'm not against having your CUDA changes in the ebuild. But I feel that what you have done is something that could help others greatly when creating ebuild with CUDA.

> I can't change it quickly and would have to be very certain that I don't have to change it as often as I might want to. I also don't have to test every ebuild affected by those changes.
I don't think that the code need to be changed that often?
It simply returns the compiler to use or what cuda arches to use.
I don't see how this could easily break ebuilds that would use this in the future if you simply just provide the path to the compiler (like you are now) and a list of arches :thinking:

Reply:
I took it to mean "I might want to change the API and I want to test it in the wild before committing to it", but I think at this point, it's time to start doing it properly given this is spreading to quite a few ebuilds now.